src/main/java/org/example/learningsystem/multitenancy/filter/TenantIdentifierFilter.java

```diff
@@ -0,0 +1,54 @@
+package org.example.learningsystem.multitenancy.filter;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.example.learningsystem.multitenancy.context.TenantInfo;
+import org.example.learningsystem.multitenancy.context.TenantContext;
+import org.example.learningsystem.multitenancy.exception.InvalidTenantIdentifierException;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+
+/**
+ * Request filter that identifies the current tenant for each request and sets the tenant context.
+ * <p>
+ * Uses the configured {@link TenantResolver} strategy to extract tenant information from the request and validates
+ * the tenant identifier format before setting the context.
+ */
+@Component
+@Slf4j
+@RequiredArgsConstructor
+public class TenantIdentifierFilter extends OncePerRequestFilter {
```

Comment:
It's also required to define order of this filter when it should be executed. It should be after `apiSecurityFilterChain` (to be able to work verified JWT). Also Spring provides `FilterRegistrationBean` to programmatically register your filter